accepted/00000-http-api.md

```diff
@@ -0,0 +1,219 @@
+// POST https://uyuni.server/rhn/manager/api/contentmanagement/update_project?label=myproject
+// {
+//   'props': {
+//     'name': 'My Project',
+//     'description': 'This is my project'
+//   }
+// }
+//
+// or
+//
+// POST https://uyuni.server/rhn/manager/api/contentmanagement/update_project
+// {
+//   'label': 'myproject',
+//   'props': {
+//     'name': 'My Project',
+//     'description': 'This is my project'
+//   }
+// }
+
+post("/contentmanagement/update_project", (req, res) -> {
+  // 1.Unwrap the session user, the query parameter named "label",
+  //   and the JSON object named "props" from the request body
+  // 2.Call to the `updateProject` method, passing the unwrapped arguments
+  // 3.Serialize and return the JSON result
+});
+```
+
+
+## 2. Authentication
+
+The web UI performs token-based authentication via a session cookie called `pxt-session-cookie`. Currently, the web UI login page exposes an API endpoint to support this process. This endpoint shall also be used for API authentication.
+
+### Login handshake demo
+
+```bash
+$ curl --include --header "Content-Type: application/json" \
+--data "{'login':'admin','password':'admin'}" \
+https://uyuni.server/rhn/manager/api/login
+
+HTTP/1.1 200 200
+...
+Set-Cookie: pxt-session-cookie=209x9fa00e2e947ace87de0743f6cf63b8b0b4bfbfca2310ce6829941fa9db947987; Max-Age=3600; Expires=Tue, 08-Feb-2022 20:05:41 GMT; Path=/; Secure; HttpOnly;HttpOnly;Secure
+...
+
+$ curl --include --cookie "pxt-session-cookie=209x9fa0..." \
+https://uyuni.server/rhn/manager/api/systems/keys
+
+HTTP/1.1 200 200
+...
+<JSON data follows>
+```
+
+
+## 3. I/O
+
+On the HTTP API side, I/O serialization is handled by the `GSON` library. However, the XMLRPC implementation does not have dedicated Java classes for the result payloads, and the serialization is defined manually to expose data in a custom structure. For this purpose, XMLRPC uses specialized "serializer" classes to write structured output on XML. These serializers must be enhanced to implement the `JsonSerializer` interface from `GSON` so that Spark can serialize the output in the same way.
+
+
+## 4. Exception Handling
+
+In XMLRPC, all the exceptions used inherit `FaultException` directly. Every exception has a unique integer code that describes an XMLRPC error. For HTTP, the `FaultException` tree must be enhanched with an HTTP status code that will be used in the response in any case of exception. Unlike the unique XMLRPC fault codes, HTTP codes will be chosen from a small subset of the standard HTTP status codes.
+
+Apart from the ehancement above, the current exception structure can be used as-is with the HTTP interface.
+
+
+## 5. Security
+
+The HTTP interface reuses the existing Spark framework. Therefore, we're typically not exposed to any new security issues. However, with the introduction of the new API, existing interface will be more critical for external use, so it would be beneficial to propose an approach to security testing.
```

Comment:
This sentence is so **strong**. Why using Spark keeps us away of new security issues? This is a dangerous assumption from a security point of view

Reply:
I might be mistaken, but I read the sentence to mean that we won't be adding any additional (as in new) attack surface compared to right now since it uses a framework that's already exposed. But feel free to correct me if I misunderstood @cbbayburt.

Reply:
Ok, I read the sentence as "since we are using Spark, we won't suffer by any new security issue in the future". I would suggest rephrasing in "The HTTP interface reuses the existing Spark framework. Therefore, the attack scenario is unchanged"